### PR TITLE
add Reader.Append method

### PR DESCRIPTION
**Pull request by rogpeppe**
This allows us to change an entry in a zip archive without
changing the original archive.

### CODE REVIEW

File: example_test.go [hunk 1]
```diff
@@ -5,10 +5,10 @@
 package zip_test
 
 import (
-	"github.com/juju/zip"
 	"bytes"
 	"compress/flate"
 	"fmt"
+	"github.com/juju/zip"
```
Comment (frankban) [inline example_test.go]: Shouldn't this be in another import group?

File: writer.go [hunk 2]
```diff
@@ -45,6 +46,25 @@ func (w *Writer) SetOffset(n int64) {
 	w.cw.count = n
 }
 
+func newAppendingWriter(r *Reader, fw io.Writer) *Writer {
```
Comment (mhilton) [inline writer.go:49]: a docstring might be helpful here.

File: reader.go [hunk 3]
```diff
@@ -181,6 +183,15 @@ func (f *File) Open() (rc io.ReadCloser, err error) {
 	return
 }
 
+// Append appends entries to the existing zip archive represented by r.
```
Comment (frankban) [inline reader.go]: represented by z?

**Comment by mhilton** [comment]
:+1: I've no reason to think it wouldn't work, but I'd like to see a test that appends the same file twice.

File: writer.go (hunk 2 above)
Comment (frankban) [inline writer.go:49]: +1

**Comment by frankban** [comment]
:+1: nice!

File: example_test.go (hunk 1 above)
Comment (rogpeppe) [inline example_test.go]: done

File: writer.go (hunk 2 above)
Comment (rogpeppe) [inline writer.go:49]: done

File: reader.go (hunk 3 above)
Comment (rogpeppe) [inline reader.go]: done

**Comment by rogpeppe** [comment]
> I've no reason to think it wouldn't work, but I'd like to see a test that appends the same file twice.

I'm not quite sure what this would be testing. It's not something that the original
tests tested. Files are just appended sequentially, so the main thing we care about
for the purposes of this new functionality is that the old index entry is elided, I think,
which is what we test for. Or are you suggesting using Append on a file that's already
the result of an Append?

**Pull request merged**